msgp/read.go

```diff
@@ -759,6 +759,45 @@ func (m *Reader) ReadBytes(scratch []byte) (b []byte, err error) {
 	return
 }
 
+// ReadBytesHeader reads the size header
+// of a MessagePack 'bin' object. The user
+// is responsible for dealing with the next
+// 'sz' bytes from the reader in an application-specific
+// way.
+func (m *Reader) ReadBytesHeader() (sz uint32, err error) {
+	var p []byte
+	p, err = m.r.Peek(1)
+	if err != nil {
+		return
+	}
+	switch p[0] {
+	case mbin8:
+		p, err = m.r.Next(2)
```

Comment:
If you want to recover from some point in history, you need to keep all of the bytes between that point and the present around in a buffer somewhere. That opens up the possibility for some catastrophic worst case performance, so the "right" solution here is going to be application-dependent. (For example: I've got an application that streams `bin` objects with sizes on the order of tens of MBs. Re-starting a stream 10MB in the past would be catastrophically expensive.)

I'm anticipating that people will use this feature something like the following:

```go
func isFatal(err error) bool {  return err != io.EOF } 

func ReadLargeFlakyBin(sink io.Writer, src *msgp.Reader) error {
   var sz uint32
   var err error
   for {
      sz, err = src.ReadBinHeader()
      if err != nil {
          if isFatal(err) {
              return err
          }
          continue
      }
      break
   }
   tot := int64(sz)
   var n int64
   for tot > 0 {
       n, err = io.CopyN(sink, src, tot)
       tot -= n
       if err != nil && isFatal(err) {
           return err
       }
   }
   return nil
}
```
In the example above, the total memory use is bounded by whatever `sink` buffers in memory.